### ISSUE TITLE
Type guards / excess object literals property checking fails with Object.assign()

### ISSUE THREAD

**Issue by 18steps** [Microsoft/TypeScript#15963]
Object spread is fixed in the nightly build, as per #13878 last commented on 2 days ago, but Object.assign() is not fixed, and it doesn't seem to be mentioned in any existing issues.

**TypeScript Version:**  2.3.2, 2.4.0-dev.20170519

**Code**

```ts
interface Limited {
  property: string;
}

const assignment = (): Limited => {
  return {
    property: 'this is fine'
  };
};

const assignment2 = (): Limited => {
  return {
    property: 'this is fine',
    property2: 'this is not fine, and shouldn\'t be',
  };
};

const nothing = {};
const somethingElse = {
  something: 'something'
};
const existing = {
  property: 'existing'
};

const assignment3 = (): Limited => {
  return {
    ...nothing,
    property2: 'this is not fine, and shouldn\'t be',
  };
};

const assignment4 = (): Limited => {
  return {
    ...somethingElse,
    property2: 'this is not fine, and shouldn\'t be',
  };
};

// - fails in 2.3.2 but not 2.4.0-dev.20170519
const assignment5 = (): Limited => {
  return {
    ...existing,
    property2: 'this should not be fine, and is not in 2.4.0-dev.20170519',
  };
};

// !
const assignment6 = (): Limited => {
  return Object.assign(existing, {
    property2: 'this should not be fine either, but is',
  });
};
```

**Expected behavior:**
Type guards should produce errors when attempting to assign a property not defined in the interface.

assignment5() and assignment6() should produce errors.

**Actual behavior:**
With Object.assign() <del>or the Object spread operator</del>, type guards fail to produce an error when attempting to assign a property not defined in the interface, as long as the required properties exist.

<del>assignment5() and</del> assignment6() does not produce an error.

**Comment by mhegazy** [comment]
this is tracked by https://github.com/Microsoft/TypeScript/issues/11100

> Object spread is fixed in the nightly build, as per #13878 last commented on 2 days ago, but Object.assign() is not fixed, and it doesn't seem to be mentioned in any existing issues.

#13878 is a different issue than what you are reporting.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
